Proposal to add Framework specific inbox source generators (#181)
dotnet/designs: accepted/2021/InboxSourceGenerators.md

> # Inbox Source Generators

Reviewer: Hi
I was wondering what Inbox Source Generators is referring to?

Reviewer 2: "Inbox" meaning they ship with the .NET runtime. For example, System.Text.RegularExpressions ships "inbox" in the .NET runtime. If/when we build a source generator for Regex, that source generator would ship along side the System.Text.RegularExpressions library in the .NET runtime.

So "Inbox Source Generators" refers to source generators that ship along with the .NET runtime.

Author: I believe this should be sufficiently defined below in scenarios.  I tried some other titles and they all seemed too generic and  implied more than just inbox.  Any suggestions for a different title that doesn't use jargon up front?

**PM** [Immo Landwerth](https://github.com/terrajobst) |
**Dev** [Eric StJohn](https://github.com/ericstj)

> Roslyn source generators allow for component developers to provide functionality that runs in the context of compilation, analyzes a users code, and can generate additional code to add to the compilation.  Compelling scenarios for source generators involve moving dynamic runtime code to be specialized and static emitted by the compiler: specializing serializer code for a target type, pre-compiling regular expressions at design time, pre-computing composition graphs for DI systems, and many more.  This techinque was possible before but was rather clunky and had to be done outside the compiler, so its use was minimal and not part of the default experience.  Roslyn source generators solve this and as a result we wish to develop many source generators which are part of the default experience for .NET applications.  Source generators present a new type of challenge for a design time component: they need to be part of the toolchain but are tied to a specific compoents within a target framework, and often re-implement or share the functionality of those components.  We need to define how we ship these source generators and make them available to developers.  We need to define the lifecycle for these source generators and how it is tied to the framework component it extends.

Reviewer 3: ```suggestion
Roslyn source generators allow for component developers to provide functionality that runs in the context of compilation, analyzes a users code, and can generate additional code to add to the compilation.  Compelling scenarios for source generators involve moving dynamic runtime code to be specialized and static emitted by the compiler: specializing serializer code for a target type, pre-compiling regular expressions at design time, pre-computing composition graphs for DI systems, and many more.  This techinque was possible before but was rather clunky and had to be done outside the compiler, so its use was minimal and not part of the default experience.  Roslyn source generators solve this and as a result we wish to develop many source generators which are part of the default experience for .NET applications.  Source generators present a new type of challenge for a design time component: they need to be part of the toolchain but are tied to a specific components within a target framework, and often re-implement or share the functionality of those components.  We need to define how we ship these source generators and make them available to developers.  We need to define the lifecycle for these source generators and how it is tied to the framework component it extends.
```

> Roslyn source generators allow for component developers to provide functionality that runs in the context of compilation, analyzes a users code, and can generate additional code to add to the compilation.  Compelling scenarios for source generators involve moving dynamic runtime code to be specialized and static emitted by the compiler: specializing serializer code for a target type, pre-compiling regular expressions at design time, pre-computing composition graphs for DI systems, and many more.  This techinque was possible before but was rather clunky and had to be done outside the compiler, so its use was minimal and not part of the default experience.  Roslyn source generators solve this and as a result we wish to develop many source generators which are part of the default experience for .NET applications.  Source generators present a new type of challenge for a design time component: they need to be part of the toolchain but are tied to a specific components within a target framework, and often re-implement or share the functionality of those components.  We need to define how we ship these source generators and make them available to developers.  We need to define the lifecycle for these source generators and how it is tied to the framework component it extends.

Reviewer 4: `a users`→`a user’s`
`techinque`→`technique`

## Scenarios and User Experience

A developer can use source generators for framework API without making any changes to their project.  Creating a new project should be sufficient.

A developer can use source generators for components in NuGet packages by referencing the NuGet package for that component.  

> When a component ships both in a NuGet package and a framework, the developer has a consistent experience with the source generator as they do with the library itself.  If the package referenced is newer then the framework the source generator from the package will be used as is already the case for the library in that package.  Similarly if the framework is newer than the NuGet package the framework's source generator will be used as is already the case for the library in the framework.

Reviewer 4: `newer then`→`newer than`

> A developer can recieve updates to a source generator through normal channels which they already obtain updates today.

Reviewer 4: `recieve`→`receive`

> A developer can update their version of Visual Studio or Dotnet SDK and have confidence that new versions of the toolchain won't impact their applciations which target older frameworks, outside of known and published servicing updates.

Reviewer 3: ```suggestion
A developer can update their version of Visual Studio or Dotnet SDK and have confidence that new versions of the toolchain won't impact their applications which target older frameworks, outside of known and published servicing updates.
```

> A developer has confidence that their library using source generators will continue to work in the same way a library without source generators works on future versions of .NET following the same compatibility rules.

Author: Library authors using inbox source generators.  I'll reword it.

## Requirements

### Goals

- Enable consumption of source generators for components in shared frameworks
- Address conflict resolution between source generators in shared frameworks and nuget packages
- Define servicing process for shared framework source generators
- Establish guidelines building and versioning shared framework source generators
- Describe compatibility requirements for source generators.

### Non-Goals

- Define a process for components not in shared frameworks to contribute source generators.
- Define any changes to the source generator architecture.
- Define any implementation details of source generators or source generator tests.
- Define a process around internal or non-shipping source generators

## Stakeholders and Reviewers

- ASP.NET - also building shared framework source generators.

> - .NET Libraries team - building multiple shared framework source generators.

Reviewer 5: I guess the CLR/diagnostics team maybe listed here for their EventSource generator.

- Mono team - interested in providing source generators.
- Roslyn team - responsible for source generator infrastructure.
- .NET SDK team - responsible for resolving shared frameworks and doing conflict resolution with content.
- NuGet team - FYI only, no changes requested

## Design

### Framework provided source generators

> Framework provided source generators should have a mechanism to be exposed and versioned along side the framework and library which they extend.  These source generators will be passed to the compiler when that framework is targeted.  These source generators will not ne used when the framework is not targeted.  For example, ASP.NET specific source generators will not be used with ASP.NET is not referenced.  The version of the analyzers specific to the TargetFramework version will be used.  For example, when targeting .NET 6.0 the analyzers specific to net6.0 will be used and not those for .NET 7.0, even when using the SDK that ships with .NET 7.0.

Reviewer 3: ```suggestion
Framework provided source generators should have a mechanism to be exposed and versioned along side the framework and library which they extend.  These source generators will be passed to the compiler when that framework is targeted.  These source generators will not be used when the framework is not targeted.  For example, ASP.NET specific source generators will not be used with ASP.NET is not referenced.  The version of the analyzers specific to the TargetFramework version will be used.  For example, when targeting .NET 6.0 the analyzers specific to net6.0 will be used and not those for .NET 7.0, even when using the SDK that ships with .NET 7.0.
```

> Framework provided source generators should have a mechanism to be exposed and versioned along side the framework and library which they extend.  These source generators will be passed to the compiler when that framework is targeted.  These source generators will not ne used when the framework is not targeted.  For example, ASP.NET specific source generators will not be used with ASP.NET is not referenced.  The version of the analyzers specific to the TargetFramework version will be used.  For example, when targeting .NET 6.0 the analyzers specific to net6.0 will be used and not those for .NET 7.0, even when using the SDK that ships with .NET 7.0.

Reviewer 3: ```suggestion
Framework provided source generators should have a mechanism to be exposed and versioned along side the framework and library which they extend.  These source generators will be passed to the compiler when that framework is targeted.  These source generators will not ne used when the framework is not targeted.  For example, ASP.NET specific source generators will not be used when ASP.NET is not referenced.  The version of the analyzers specific to the TargetFramework version will be used.  For example, when targeting .NET 6.0 the analyzers specific to net6.0 will be used and not those for .NET 7.0, even when using the SDK that ships with .NET 7.0.
```

> Framework provided source generators should have a mechanism to be exposed and versioned along side the framework and library which they extend.  These source generators will be passed to the compiler when that framework is targeted.  These source generators will not ne used when the framework is not targeted.  For example, ASP.NET specific source generators will not be used when ASP.NET is not referenced.  The version of the analyzers specific to the TargetFramework version will be used.  For example, when targeting .NET 6.0 the analyzers specific to net6.0 will be used and not those for .NET 7.0, even when using the SDK that ships with .NET 7.0.

Reviewer 6: How are the source generators going to interact with optional workloads?  I'm concern about ref pack dependency which is not known for some of the optional workloads from a runtime perspective.

Reviewer 7: Optional workloads can provide their own targeting packs, so they could also provide source generators to go along with them.  They would need to follow the same restrictions about the dependencies they can take.  Is that what your concern is?

Reviewer 6: No, what I had in mind is that optional workloads could share some part of tooling (e.g. AOT toolchain) which might require source generators. Such code generators could be workload or runtime pack specific.

Author: Today runtimes don't leak into the compile phase.  I think if you start doing that then you really should have a different TFM, or we should be adding support for cross-compilation by runtime cc @terrajobst

Author: You can see a bit of this in the Q & A section titled 
> Why in the ref pack, why not runtime pack?

> Framework provided source generators should have a mechanism to be exposed and versioned along side the framework and library which they extend.  These source generators will be passed to the compiler when that framework is targeted.  These source generators will not ne used when the framework is not targeted.  For example, ASP.NET specific source generators will not be used when ASP.NET is not referenced.  The version of the analyzers specific to the TargetFramework version will be used.  For example, when targeting .NET 6.0 the analyzers specific to net6.0 will be used and not those for .NET 7.0, even when using the SDK that ships with .NET 7.0.

Reviewer 4: `ne`→`be`

> To facilitate this, we will package analyzers inside [reference packs](https://github.com/dotnet/designs/blob/main/accepted/2019/targeting-packs-and-runtime-packs.md).  We will use the standard [nuget conventions for describing analyzers](https://github.com/dotnet/designs/blob/main/accepted/2019/targeting-packs-and-runtime-packs.md).  The .NET SDK will be responsible for locating the appropriate analyzers for the TargetFramework and creating `@(Analyzer)` items as part of the `ResolveTargetingPackAssets` task.  These items will be conflict-resolved with other Analyzer sources based as described in the following section.

Reviewer 3: It is very rare for reference packs to need servicing today. Source generators are much more likely to contain bugs that require servicing. Should it be mentioned here? Are there any challenges that this is going to create?

Author: I cover this below in Q/A.  IMHO we ship *a* ref pack regardless, similar to how we ship the runtime regardless, so no matter what we are shipping the same number of bits.  We also need to make sure the ref-pack is serviceable in general.  We shouldn't let a potential bug in something that needs to work lead us down a different path.

If it weren't the ref-pack then it would be something else, in which case we'd have even more moving pieces.

> To facilitate this, we will package analyzers inside [reference packs](https://github.com/dotnet/designs/blob/main/accepted/2019/targeting-packs-and-runtime-packs.md).  We will use the standard [nuget conventions for describing analyzers](https://github.com/dotnet/designs/blob/main/accepted/2019/targeting-packs-and-runtime-packs.md).  The .NET SDK will be responsible for locating the appropriate analyzers for the TargetFramework and creating `@(Analyzer)` items as part of the `ResolveTargetingPackAssets` task.  These items will be conflict-resolved with other Analyzer sources based as described in the following section.

Reviewer 7: We don't use the NuGet conventions to interpret the contents of targeting packs.  The contents of a targeting pack are described by a `data/FrameworkList.xml` file.  The paths that the files are actually found under may (or may not) conform to the NuGet conventions.

So we will need to define how analyzers are included in targeting packs.  I would go with either adding them to the FrameworkList.xml file, or possibly creating a new file specific for analyzers.

Author: We do already follow nuget convention for refs.  I would prefer we follow the same convention for analyzers, even if the actual resolution is more of a simple path check.  That way we don't "forget" about some constraining variable like language.

I thought we used a wild-card for refs but now I see that the SDK constructs all the paths from information in the framework list.
https://github.com/dotnet/sdk/blob/8659ce5f7550359048083a1ddafb25f105d269b8/src/Tasks/Microsoft.NET.Build.Tasks/ResolveTargetingPackAssets.cs#L58

Is this a perf optimization to avoid hitting the disk?

Author: Ok, I think that's the case.  I'll address this in the doc.

Reviewer 7: Yes, it's partly about perf.  I'm not sure enumerating the files on disk would impact perf, but we also store the assembly and file versions in the framework list which helps.

It's also helpful to rely on the framework list for the file path instead of conventions, so we don't have to worry about whether we should use net5.0 or net5.0-windows (or netcoreapp5.0) as part of the path, for example.  We already used the target framework to select the targeting pack, so using it as part of a convention for the path inside the pack just complicates things.

For analyzers, rather than re-implement the NuGet convention (which actually [isn't currently implemented in NuGet](https://github.com/NuGet/Home/issues/6279)), I think it's better to just have them listed in a file such as the framework list.

> To facilitate this, we will package analyzers inside [reference packs](https://github.com/dotnet/designs/blob/main/accepted/2019/targeting-packs-and-runtime-packs.md).  We will use the standard [nuget conventions for describing analyzers](https://github.com/dotnet/designs/blob/main/accepted/2019/targeting-packs-and-runtime-packs.md).  The .NET SDK will be responsible for locating the appropriate analyzers for the TargetFramework and creating `@(Analyzer)` items as part of the `ResolveTargetingPackAssets` task.  These items will be conflict-resolved with other Analyzer sources based as described in the following section.

Reviewer 8: These paragraphs talk about analyzers... I assume that was meant to be inclusive of source generators?

Author: I'll clarify.

### Source generator conflict resolution

Source generators and analyzers are passed to the compiler as @(Analyzer) items.  These items may come from NuGet packages, be built into the SDK targets, be directly defined by the user via ProjectReference or raw Analyzer items, or be provided by the new framework mechanism described in this document.  All should be considered for conflict resolution based on the same constraint.

> The .NET SDK already does conflict resolution for other asset types: references and copy-local runtime assets.  It should do the same for Analyzers.  Analyzers need not have "platform manifest" inputs for conflict resolution since all analyzers will be available during build, unlike runtime files which may not be available.  Comparison for the sake of conflict resolution can consider the same assembly and file version rules as reference files.  When finding two or more source generators with the same name

Reviewer: > When finding two or more source generators with the same name

Is this an unfinished sentence?

Source generator conflict resolution should run before compile, but after all NuGet package assets have been evaluated.  Specific sequencing, public/private target naming, extensibility will not be specified here but should follow best practice as determined by the SDK team. 

### Source generator versioning

> Source generators should be strongly named and update assembly version with every public stable release (GA and servicing).  This ensures that the publicly shipped source generators can be loaded side-by-side in the same process and garuntees 

Reviewer 3: Unfinished sentence?

Author: Hmm, that must have been lost.  Need to see if I have in a previous version.

### Source generator compatibility concerns

> Source generators should only use public API in framework components they extend.  This public API already needs to meet the strict binary compatibility garuntees as framework API.  As such libraries containing source-generated code will be binary-compatible with future releases, just as if the user wrote this code directly.

Reviewer 3: ```suggestion
Source generators should only use public API in framework components they extend.  This public API already needs to meet the strict binary compatibility guarantees as framework API.  As such libraries containing source-generated code will be binary-compatible with future releases, just as if the user wrote this code directly.
```

> Source generators should only use public API in framework components they extend.  This public API already needs to meet the strict binary compatibility guarantees as framework API.  As such libraries containing source-generated code will be binary-compatible with future releases, just as if the user wrote this code directly.

Reviewer 8: > Source generators should only use public API in framework components they extend

But should be able to build source used by the framework component into the source generator as well.  For example, regex's parser is non-public API and we can build the source for it into both Regex and the regex source generator.  This doesn't create a dependency between them any more than implementing a custom parser for the source generator does.

Author: Right I think I cover part of this when I talk about source sharing.  I'll call out how source sharing may be a strategy for avoiding public API.

Author: Actually this section is talking about the API used by the generated code.  I'll make sure that's clearer.

> Source generators should not use public API with "reserved" or "undocumented" functionality in attempts to avoid public API garuntees.  Libraries paired with source generators need to provide the same garuntees as libraries alone do today.

Reviewer 3: ```suggestion
Source generators should not use public API with "reserved" or "undocumented" functionality in attempts to avoid public API guarantees.  Libraries paired with source generators need to provide the same guarantees as libraries alone do today.
```

> Source generators should avoid emitting public API in user's projects.  Such API can create a compatibility contract in that user's library which needs to be maintained.  If a source generator produces public API in a user's project by design, that public API must be strictly maintained following the same rules as framework components across versions of the source generator.  Care should be taken to the public API which is generated and should be reviewed by the Framework Design Council.  A similar constraint applies to internal API: users will have source that expects to interact with the generated internal API and the user will expect that code to continue to compile between versions.

Reviewer 6: > Source generators should avoid emitting public API in user's projects.

Do we have analyzers for that ;-)

Author: No, just guidelines.  Sort of like the Fx design guidelines.  I suspect we'll learn a lot more here over time.  Maybe a reason for @bartonjs to add another chapter to that book :)

> Source generators should avoid emitting public API in user's projects.  Such API can create a compatibility contract in that user's library which needs to be maintained.  If a source generator produces public API in a user's project by design, that public API must be strictly maintained following the same rules as framework components across versions of the source generator.  Care should be taken to the public API which is generated and should be reviewed by the Framework Design Council.  A similar constraint applies to internal API: users will have source that expects to interact with the generated internal API and the user will expect that code to continue to compile between versions.

Reviewer 4: `in user's projects`→`in a user's projects` as in other places
or
`in user's projects`→`in users' projects`

> Generated code should not be generated in System namespaces, it should prefer the user's namespace.  Generated code should be triple-slashed commented and be free from compiler warnings.  Generated code should avoid name colisions with user code and provide affordances for resolving name conflicts. 

Reviewer 6: Is there any general guidance on how the generated code should look like (e.g. name prefix, special attributes, version suffix, etc)?

Reviewer 9: Have you thought about using other version specific features such as nullability? Razor side-steps documentation and  nullability by pragma suppressing them.

Reviewer 5: There might be a debuggability connection here, eg., if the code is marked `<autogenerated />`

In general I have not seen what the debuggability story is for source generators. If you are debugging on the same machine you generated on, I guess the PDB points to a generated file in temp (which I believe is generated only for this purpose). But what is the story if you are debugging on a different machine? Is the source serialized into the PDB? I have not seen details of this, perhaps you have a link.

Reviewer 5: cc @cartermp

Author: I don't intend to go into super detail around strategies here.  If possible we should like to design considerations the compiler takes when doing similar things.  I guess the unique problem here is that the compiler usually generates code that it doesn't expect callers to interact with vs source generators specifically generate code that the user needs to interact with.

Author: > Is the source serialized into the PDB?

Yes, it should be.  You can also inspect it in the IDE, or even have the compiler write to disk with `EmitCompilerGeneratedFiles`.  I couldn't find the docs on this, @chsienki or @jasonmalinowski might know.

> Have you thought about using other version specific features such as nullability? Razor side-steps documentation and nullability by pragma suppressing them.

We should annotate the generated code so that it works well with user's types, be they nullable or not.  I will add a note.  In general we should be trying to write the *best code possible* rather than coming up with reasons for not following the rules, except maybe for tabs vs spaces.  That rule should of course be suppressed ;)

> Generated code should not be generated in System namespaces, it should prefer the user's namespace.  Generated code should be triple-slashed commented and be free from compiler warnings.  Generated code should avoid name colisions with user code and provide affordances for resolving name conflicts. 

Reviewer 4: `colisions`→`collisions`

### Functionality concerns for source generators

Though the following are not specific to inbox source generators, these concerns are worth reviewing to describe how they apply to framework specific source generators.

> Source generators target .NETStandard and must run on both .NETFramework and .NETCore.  

Reviewer 6: Would that be Analyzer Nuget packing error if that's not the case?

Author: Not sure if there's any nuget validation for this today.  @cartermp might know.  

Similar requirements are present for MSBuild tasks already and I don't think we enforce it anyhwere.  cc @dsplaisted 

I expect for our source-generators in dotnet/runtime to have generator unit tests which run the generator in both environments as @layomia has set up already for Json: https://github.com/dotnet/runtimelab/blob/2faaae33587a5a53e8b948d1f9e2b6b044e9bc37/src/libraries/System.Text.Json/System.Text.Json.SourceGeneration.UnitTests/System.Text.Json.SourceGeneration.UnitTests.csproj#L3

Author: I think I misread your question: 
> Would that be Analyzer Nuget packing error if that's not the case?

It would be an error in *building* the analyzer and simple functional testing of the analyzer from visual studio would fail if the analyzer targeted .NET core.  Analyzers can really only be packaged in [one way](https://docs.microsoft.com/en-us/nuget/guides/analyzers-conventions) today: there is no cross-targeting.  This document is laying out a new way for frameworks.  Still there is only one way to package it and have it be used.

> Source generators target .NETStandard and must run on both .NETFramework and .NETCore.  

Reviewer 8: > Source generators target .NETStandard

Meaning the code in the source generator itself, not the code it generates.

Author: Correct.  Will clarify.

Reviewer 5: When determining what the target runtimes are of the generated code, the generator examines the types available in the references passed in, right? As the compiler does to light up some features. It doesn't look at the TFM.

Reviewer 3: It is up to the each source generator whether to light-up or not. I expect that most of the inbox source generators will be hardcoded to latest runtime version, and won't do any light up.

> Source generators should limit their dependencies.  There is no scheme for encoding framework-specific nor runtime-specific implementations of dependencies or source generators based on the runtime environment so source generators should avoid any dependencies on assemblies which need to differ by framework or runtime.  Source generators may depend on packages which are known dependencies of the compiler and should depend on versions less than or equal to that provided by the compiler and should not include that dependency in their package.

Reviewer 9: > There is no scheme for encoding framework-specific nor runtime-specific implementations of dependencies or source generators based on the runtime environment

This doesn't seem like a particularly difficult problem for the SDK to solve if it needed to, no?

Reviewer 7: It's actually probably pretty hard.  And source generators that are delivered as NuGet packages don't support these types of dependencies, so we shouldn't try to add them to in-box source generators either (unless / until the support for NuGet source generators changes).

Author: Right: I'm specifically calling this out as hard and should be avoided.  I wanted to call out that "provided" packages can be used, like System.Reflection.Metadata, System.Collections.Immutable, System.Memory.

For this problem to be "solved" you need the host to solve it, so roslyn would need to have a more complicated plugin model involving something like deps files or a custom ALC, dll.configs, and something brand new that can multiplex between frameworks at runtime.  Let's not go there.

> Source generators should not directly reference nor execute runtime code.  Despite inbox source generators being coupled to a specific framework version, source generators cannot assume they will be running on that framework.  They may be running on a newer, older, or completely different framework.  There is no type-equivalence garunteed between references of the source generator and those of the user code that is under analysis.  As such source generators will need to examine user code references by name, or when needing to make equivalence checks can resolve types in the user code's type reference set to ensure valid equivalence checks.  Source sharing should be considered as an alternative for cases where source generators need to execute runtime library code.

Reviewer 3: ```suggestion
Source generators should not directly reference nor execute runtime code.  Despite inbox source generators being coupled to a specific framework version, source generators cannot assume they will be running on that framework.  They may be running on a newer, older, or completely different framework.  There is no type-equivalence guaranteed between references of the source generator and those of the user code that is under analysis.  As such source generators will need to examine user code references by name, or when needing to make equivalence checks can resolve types in the user code's type reference set to ensure valid equivalence checks.  Source sharing should be considered as an alternative for cases where source generators need to execute runtime library code.
```

> Source generators should not directly reference nor execute runtime code.  Despite inbox source generators being coupled to a specific framework version, source generators cannot assume they will be running on that framework.  They may be running on a newer, older, or completely different framework.  There is no type-equivalence garunteed between references of the source generator and those of the user code that is under analysis.  As such source generators will need to examine user code references by name, or when needing to make equivalence checks can resolve types in the user code's type reference set to ensure valid equivalence checks.  Source sharing should be considered as an alternative for cases where source generators need to execute runtime library code.

Reviewer 3: Should we also have chapter on Security concerns for source generators?

The security bugs in code generated by source generators cannot by serviced by traditional runtime servicing channel. They require rebuilding with the serviced toolchain. 

- The exact version of the source generator used to build the assembly needs to be recorded somewhere (where?), so that it is possible to flag the assembly as being built with vulnerable version of the source generator.

- The code generated by the source generator should aim to have low security risk. For example, the source generators should not be generating unsafe buffer manipulation operations inline. Instead, the source generators should prefer to call APIs in the framework to perform unsafe buffer manipulations where possible.

Author: Initially this doc was only talking about distribution but we might as well grow it to discuss this as well.  It's not a bad idea to preserve some metadata about the generator for identification purposes.  Should we add assembly metadata?  We can't offer any automatic updates of course.

Reviewer 3: Assembly metadata and/or record the source generator details in PDBs are the two options that I can think of.

Author: I have a few more, will update the doc.

## Q & A

### Why not just deliver source generators in the SDK same as roslyn analyzers?

Doing so would require source generators to carefully version and test across all framework versions.  This would increase testing cost for source generators and increase risk when shipping updates to source generators since they would apply to past, stable TargetFrameworks.  Conceptually source generators are lifting a substantial amount of runtime functionality out of a library and executing that at design time. This code is directly included in the user's application and contributes to runtime behavior.

### Inplace servicing updates are good enough for the compiler, why not source generators?

The compiler has a very stable 1:1 relationship between a user's source and the code it generates. As such the user has the ability to control the output of the compiler through their own usage.  The compiler's contribution of complexity to the user's assembly is very small and highly specified.
 

> Contrast this to source generators which can take a very small amount of user source to produce a large amount of generated code.  This large amount of generated code contains significant complexity which is loosely specified.  It's behavior is specific to the component it extends.  It's likely that we'd need to bugfix this code over time, and change it as we develop features in the components it extends.  It's also likely that we will learn new techniques and improvments to source generators over time, increasing the changes for bugfixes and features

Reviewer 4: `It's behavior`→`Its behavior`

Missing period at the end of the sentence

> User's expect stability for projects which do not change, for example a project targeting an LTS release. It is much more difficult to ensure this stability when we need to significantly components which contribute complexity to that project.

Reviewer 4: `User's expect`→`Users expect`

### Is it bad to have multiple versions of a source generator possibly loaded by a repository?

A repository may have projects targeting multiple frameworks.  These frameworks may contain different versions of the same source generator.  When building the repository this may require the same compiler server instance to load multiple copies of a source generator.  The same compiler would not likely load multiple copies of analyzers which do not ship side by side.  

### Should all inbox source generators be tied to a shared framework?

No, some source generators might be independent of libraries in a shared framework: these could go in a non-side-by-side location (eg: analyzers folder).  Other source generators may not contain functionality, but may instead just build some convenience API on top of existing public API.  These could be considered stable enough to place in a non-side-by-side location.

### Should only source generators be shipped this way?  What about analyzers and code-fixes?

> Nothing is preventing analyzers and code fixes from shipping in the same manner.  If it's desireable for a such a component to ship with the framework rather than the compiler then that component may ship in this manner.

Reviewer 4: `desireable`→`desirable`

### Why in the ref pack, what not some other "analyzer" pack?

> The SDK already has a means for acquisition and selection of ref packs.  The inputs to the selection of ref packs are the same as that of source generators (Version, FrameworkReference, and TargetFramework).  Ref packs user normal nuget convention which support analyzers and source generators.

Reviewer 7: They don't use normal NuGet conventions and won't automatically support source generators / analyzers.  We'll have to implement this in the .NET SDK.

Author: Right, I expect that.  I'll call it out.

> The SDK already has a means for acquisition and selection of ref packs.  The inputs to the selection of ref packs are the same as that of source generators (Version, FrameworkReference, and TargetFramework).  Ref packs user normal nuget convention which support analyzers and source generators.

Reviewer 4: `nuget`→`NuGet`

> It may be desireable to create a seperate package just for analyzers in order to reduce risk in servicing.  We support servicing ref packs today, but we have to be careful not to expose new API nor change any assembly identites.  

Reviewer 3: ```suggestion
It may be desirable to create a separate package just for analyzers in order to reduce risk in servicing.  We support servicing ref packs today, but we have to be careful not to expose new API nor change any assembly identities.  
```

### Why in the ref pack, what not runtime pack?

> Runtime packs are currently independent of compile.  Sourcing inputs for compile from a runtime-specific package would break this and require signicant changes to the user experience, such as permitting cross-compiling by runtime or "targeting" a runtime.

Reviewer 4: `signicant`→`significant`

###  I really really need to execute runtime code in the TargetFramework!

> You probably need a designer or indepent build component that has a host process that can leverage the runtime targeting information of the application.  At the moment this doesn't have a great pattern to follow with source generators.  The compiler does not provide a host, nor does it know anything about target runtime (or in some cases executable frameworks).

Reviewer 4: `indepent`→`independent`

## Scenarios and User Experience

A developer can use source generators for framework API without making any changes to their project.  Creating a new project should be sufficient.

A developer can use source generators for components in NuGet packages by referencing the NuGet package for that component.  

When a component ships both in a NuGet package and a framework, the developer has a consistent experience with the source generator as they do with the library itself.  If the package referenced is newer than the framework the source generator from the package will be used as is already the case for the library in that package.  Similarly, if the framework is newer than the NuGet package the framework's source generator will be used as is already the case for the library in the framework.

A developer can receive updates to a source generator through normal channels which they already obtain updates today.

A developer can update their version of Visual Studio or Dotnet SDK and have confidence that new versions of the toolchain will not impact their applications which target older frameworks, outside of known and published servicing updates.

> A developer has confidence that their library using inbox source generators will continue to work in the same way a library that does not use inbox source generators works on future versions of .NET.  Source gener

Reviewer 9: This line got cut-off

- Address conflict resolution between source generators in shared frameworks and NuGet packages
- Define servicing process for shared framework source generators
- Establish guidelines building and versioning shared framework source generators
- Describe compatibility requirements for source generators.

### Non-Goals

- Define a process for components not in shared frameworks to contribute source generators.
- Define any changes to the source generator architecture.
- Define any implementation details of source generators or source generator tests.
- Define a process around internal or non-shipping source generators

## Stakeholders and Reviewers

- ASP.NET - also building shared framework source generators.
- .NET Libraries team - building multiple shared framework source generators.
- Mono team - interested in providing source generators.
- .NET Runtime team - interested in providing source generators.
- Roslyn team - responsible for source generator infrastructure.
- .NET SDK team - responsible for resolving shared frameworks and doing conflict resolution with content.
- NuGet team - FYI only, no changes requested

> ## Design

Reviewer 10: I think we should mention what this design means for branches and the servicing process. Specifically, that if a bug is found in a generator we may need to make the same fix in multiple branches (potentially using different code), depending on how many frameworks we support and in how many frameworks the bug is present. In other words, the process would be virtually identical to how we service runtime code.

Author: I'll add to the servicing section.

### Framework provided source generators

> Framework provided source generators should have a mechanism to be exposed and versioned along side the framework and library which they extend.  These source generators will be passed to the compiler when that framework is targeted.  These source generators will not be used when the framework is not targeted.  For example, ASP.NET specific source generators will not be used when ASP.NET is not referenced.  The version of the source generators specific to the TargetFramework version will be used.  For example, when targeting .NET 6.0 the source generators specific to net6.0 will be used and not those for .NET 7.0, even when using the SDK that ships with .NET 7.0.

Author: I'll try to address this.

TFM seems like a solution here but that glosses over the fact that analyzers could be delivered out of band as well.  TFM is a correlation tool where you really want causation.  I believe the best practice is to do API-based light-up.

Source generators, analyzers, and code fixes are all represented and communicated in the same manner to the compiler, as analyzer dlls using the `-analyzer` switch.  Though this document is motivated by describing the process for source generators it will work in the same manner for analyzers and in some places we may refer to the assemblies containing source generators as analyzers due to the way they are understood by the compiler, package manager, and SDK.

> To facilitate this, we will package source generators inside [reference packs](https://github.com/dotnet/designs/blob/main/accepted/2019/targeting-packs-and-runtime-packs.md).  We will use the standard [NuGet conventions for describing analyzers](https://docs.microsoft.com/en-us/nuget/guides/analyzers-conventions) with respect to the location in the package, however the SDK will not probe this path.  All source generators and their metadata will be listed in FrameworkList.xml, similar to references.  Analyzers will be added as `File` elements with `Type="Analyzer"` (constrasting to `Type="Managed"` which is currently used for references) and optionally `Language="cs|vb"`.  Omitting the `Language` attribute means the analyzer applies to all languages.  The .NET SDK will be responsible for locating the appropriate source generators from the refpack, based on the project's `$(Language)` and creating `@(Analyzer)` items as part of the `ResolveTargetingPackAssets` task.  These items will be conflict-resolved with other `@(Analyzer)` sources as described in the following section.  When viewed from the IDE it should be clear that the source of the `Analyzer` is the framework reference, similarly to how it's clear that the source of NuGet package analyzers come from packaages.

Author: @dsplaisted please review.  I chose `<File Type="Analyzer"` but could also imagine us using `<Analyzer` if we were worried that older SDK's might ever see these files and mistreat analyzers as references.

Reviewer 7: Looks good.  We'll have to figure out with the project system team how we'll annotate the Analyzer items so that they can associate them with the right framework reference in solution explorer.

Author: I wrote this based on what I would expect without actually checking how things work today.  It's entirely possible that this doesn't work well for package-analyzers today.  I did notice that the IDE has a separate node for analyzers and lists the rules that they expose.

Alternative designs are listed in the [Q & A](#Q%20&%20A) section below.

### Source generator conflict resolution

Source generators and analyzers are passed to the compiler as @(Analyzer) items.  These items may come from NuGet packages, be built into the SDK targets, be directly defined by the user via ProjectReference or raw Analyzer items, or be provided by the new framework mechanism described in this document.  All should be considered for conflict resolution based on the same constraint.

> The .NET SDK already does conflict resolution for other asset types: references and copy-local runtime assets.  It should do the same for source generators, ore more generally Analyzers.  Analyzers need not have "platform manifest" inputs for conflict resolution since all analyzers will be available during build, unlike runtime files which may not be available.  Comparison for the sake of conflict resolution can consider the same assembly and file version rules as reference files.  When finding two or more source generators with the same name conflict resolution will select the generator with higher assembly version, then higher file version, then that which is in the framework.  This parallels the same rules followed by reference assemblies.  The SDK may choose to optimize this comparison as it does for reference assemblies, by including a data file that includes analyzer metadata.

Reviewer 9: ```suggestion
The .NET SDK already does conflict resolution for other asset types: references and copy-local runtime assets.  It should do the same for source generators, or more generally Analyzers.  Analyzers need not have "platform manifest" inputs for conflict resolution since all analyzers will be available during build, unlike runtime files which may not be available.  Comparison for the sake of conflict resolution can consider the same assembly and file version rules as reference files.  When finding two or more source generators with the same name conflict resolution will select the generator with higher assembly version, then higher file version, then that which is in the framework.  This parallels the same rules followed by reference assemblies.  The SDK may choose to optimize this comparison as it does for reference assemblies, by including a data file that includes analyzer metadata.
```

> Source generator conflict resolution should run before compile, but after all NuGet package assets have been evaluated.  Specific sequencing, public/private target naming, extensibility will not be specified here but should follow best practice as determined by the SDK team. 

Author: Correct.

Source generators should not generate code which uses public API with "reserved" or "undocumented" functionality in attempts to avoid public API guarantees.  Libraries paired with source generators need to provide the same guarantees as libraries alone do today.

Source generators should avoid emitting public API in a user's projects.  Such API can create a compatibility contract in that user's library which needs to be maintained.  If a source generator produces public API in a user's project by design, that public API must be strictly maintained following the same rules as framework components across versions of the source generator.  Care should be taken to the public API which is generated and should be reviewed by the Framework Design Council.  A similar constraint applies to internal API: users will have source that expects to interact with the generated internal API and the user will expect that code to continue to compile between versions.

> Generated code should not be generated in System namespaces, it should prefer the user's namespace.  Generated code should be triple-slashed commented and be free from compiler warnings.  Generated code should avoid name collisions with user code and provide affordances for resolving name conflicts.  As much as possible generated code should meet the quality statndards of .NET library code, using nullable annotations meeting framework design and naming guidelines, etc.  It should feel like a natural extension of the library.

Reviewer 11: "using nullable annotations meeting framework design and naming guidelines, etc"

I feel like you meant that it should use nullable annotations and also follow design guidelines.  A comma would help, or flipping the clauses

"As much as possible, the generated code should meet the quality standards of .NET library code, adhere to framework design and naming guidelines, utilize nullable annotations, etc."

In my sentence I also fixed `statndards` to `standards` :smile:

### Source generator servicing

> Source generators contribute significant code to user assemblies and that code may have issues.  We have very little control over changing customer assemblies once they have consumed a source generator.  We cannot and should not attempt to change the behavior of customer's assemblies (eg: hot patch) to fix issues.  Should an issue be found we should fix it in the source generator and ship a new version, just as we would if we found an issue with compiler.  Customer's desiring this fix must proactively acquire the new code-generator, typically through an updated SDK which carries the ref pack, rebuild their code, and update their production applications.

Reviewer 11: `Customer's` => `Customers` (plural, not possessive)

Due to the limited servicing agility for code-generated code, we should be very careful about the complexity of the code which is generate.  Where possible we should limit code-generated code to "glue code", only that which must be specific to the user assembly or types.  We should try to put more complex code inside the framework or library itself, even if it means exposing public API that is specifically for source generated code to call.

> **Open issue:** should we do anything to make it easier to identify assemblies which might need an update?  We can consider some case studies.  

Reviewer 11: ```
$"[GeneratedCodeAttribute(
    \"{GetType().AssemblyQualifiedName}\",
    \"{Assembly.ExecutingAssembly.MoreWordsHere.FileVersion}\")]"
```

perhaps?

Author: This isn't a technical question, it's a policy/experience.  The technical side of it is less interesting and should follow from policy side.

Author: I'll use your suggestion in the list of options/examples lower in the doc, it's better than AssemblyMetadataAttribute.

> If a specific compiler version has a code-gen bug it isn't clear from the final assembly if that bug is present, unless you directly inspect the IL/bytecode for the issue.  The compiler version is embedded in the PDB of the assembly which could be used to identify assemblies with the issue.  There is no notification at design time or runtime of any issue with the compiler, the user needs to subscribe to updates to understand when an update is needed.

Author: We don't have tooling that statically links the framework into libraries.  Suppose library authors use source generators in their library, some consumer of that library might want to know that the library has a bug in its source generated code?

> If a runtime assembly has a bug we have multiple avenues of detection.  We could examine the assembly metadata files contained in the application layout, including for self-contained applications.  The runtime will also write "breadcrumb" files on a machine when code is executed that is annotated for servicing that help identify that a specific assembly or package version has been executed on that machine.  Today we use these breadcrumbs to help target updates to a machine.  Those updates can install in a well known location and be preferred over the application's copy.

Author: You're talking about single-file?  I see that as a bit different as that's a decision made by the app author who does so knowing they're going to own the deployment.  When libraries are preserved there's a different expectation.

So for means of detection we have:
1. Manual
2. PDB indicator: eg source generator assembly version, file version, hash)
3. Assembly metadata in consuming user project: eg AssemblyMetadata("SourceGeneratedBy", "source generator assembly version, file version, hash")
4. Runtime breadcrumbs

For means of update we have:
1. No notification, user needs to subscribe to servicing updates.
2. Design time notification.
3. Killbit in the runtime to prevent execution of user assemblies that contain vulnerable source generated code.
4. ~~Automatically update user assembly though a hotpatch applied by runtime or JIT~~

We'll need to better understand what customer's expectations are around servicing to decide if we should do any work on these axes.  My initial inclination is to only have a PDB indicator and require users to subscribe to updates.

### Functionality concerns for source generators

Though the following are not specific to inbox source generators, these concerns are worth reviewing to describe how they apply to framework specific source generators.

> Source generator assemblies should target .NETStandard and must run on both .NETFramework and .NETCore. Source generators should support generating source for all TargetFrameworks supported by their companion library.  In the case of inbox source generators discussed in this document this need only be the TargetFramework the generator ships in.  In the case of a NuGet package provided source generator, this should be all frameworks supported by package.

Reviewer 11: Is the .NET Standard targeting requirement coming from the fact that the compiler itself is .NET Standard 2.0-based?  So even if it's a generator for System.SomeCoolThingForNet6WithNoOlderOffering.dll-based components it will be invoked by the compiler, so is bound by the compiler's requirements?

(If so, it might seem more like "this is a technical constraint to be aware of" than "this is a decision of aesthetics/politics/something that you may want to argue with" to mention that)

Author: Correct, technical constraint.

Author: Clarifying this.

> Source generator assemblies should target .NETStandard and must run on both .NETFramework and .NETCore. Source generators should support generating source for all TargetFrameworks supported by their companion library.  In the case of inbox source generators discussed in this document this need only be the TargetFramework the generator ships in.  In the case of a NuGet package provided source generator, this should be all frameworks supported by package.

Author: Yeah, this was already updated to clarify that requirement.

Source generators should limit their dependencies.  There is no scheme for encoding framework-specific nor runtime-specific implementations of dependencies or source generators based on the runtime environment so source generators should avoid any dependencies on assemblies which need to differ by framework or runtime.  Source generators may depend on packages which are known dependencies of the compiler and should depend on versions less than or equal to that provided by the compiler and should not include that dependency in their package.

> Source generators should not directly reference nor execute runtime code.  Despite inbox source generators being coupled to a specific framework version, source generators cannot assume they will be running on that framework.  They may be running on a newer, older, or completely different framework.  There is no type-equivalence guaranteed between references of the source generator and those of the user code that is under analysis.  As such source generators will need to examine user code references by name, or when needing to make equivalence checks can resolve types in the user code's type reference set to ensure valid equivalence checks.  Source sharing should be considered as an alternative for cases where source generators need to execute runtime library code.

Reviewer 11: > Source generators should not directly reference nor execute runtime code.

Maybe it's because the coffee hasn't kicked in yet, but I'm at a loss as to what this means.  The source generators are .NET assemblies, intended to make some algorithmic decisions, and act on them... seems like things like GetEnumerator() and string.Format are going to be highly utilized.

Maybe it's saying that we shouldn't be providing public API in the 'target' library to be a "generator helper", where the output of a given version of the generator could vary by the runtime version used by a particular compiler invocation.  (I can't come up with any phrasing of this concept that would enable us to write a generator for (or to avoid) Regex without using a weasel phrase like "generator helper", since the Regex source generator could potentially use Regex as a "stable platform component" as an implementation detail)

Author: This is meant to remind people that they shouldn't treat references as executable assemblies.  We have a lot of runtime developers that we're asking to write design time code and they need to understand the difference and best practice.

>  since the Regex source generator could potentially use Regex as a "stable platform component" as an implementation detail

Be very careful about that.  You'd need to be sure that the Regex in .NETFramework is sufficient for those purposes.  You'd also need to bridge between the developer's source references and the framework component.

I will try to clarify this.

## Q & A

### Why not just deliver source generators in the SDK same as roslyn analyzers?

Doing so would require source generators to carefully version and test across all framework versions.  This would increase testing cost for source generators and increase risk when shipping updates to source generators since they would apply to past, stable TargetFrameworks.  Conceptually source generators are lifting a substantial amount of runtime functionality out of a library and executing that at design time. This code is directly included in the user's application and contributes to runtime behavior.

### Inplace servicing updates are good enough for the compiler, why not source generators?

> The compiler has a very stable 1:1 relationship between a user's source and the code it generates. As such the user has the ability to control the output of the compiler through their own usage.  The compiler's contribution of complexity to the user's assembly is very small and highly specified.

Author: I understand the compiler does code-spit for these but I still consider small in comparison to what source generators do and are capable of doing.  I did consider async, lambda, dynamic etc when writing this and that's why I used terms like "complexity" and not code size.

### Is it bad to have multiple versions of a source generator possibly loaded by a repository?

A repository may have projects targeting multiple frameworks.  These frameworks may contain different versions of the same source generator.  When building the repository this may require the same compiler server instance to load multiple copies of a source generator.  The same compiler would not likely load multiple copies of analyzers which do not ship side by side.  

### Should all inbox source generators be tied to a shared framework?

No, some source generators might be independent of libraries in a shared framework: these could go in a non-side-by-side location (eg: analyzers folder).  Other source generators may not contain functionality, but may instead just build some convenience API on top of existing public API.  These could be considered stable enough to place in a non-side-by-side location.

### Should only source generators be shipped this way?  What about analyzers and code-fixes?

> Nothing is preventing analyzers and code fixes from shipping in the same manner.  If it's desirable for a such a component to ship with the framework rather than the compiler then that component may ship in this manner.

Reviewer 10: I'd say we should strive to ship analyzers how we do today and only ship analyzers like source generators if the analyzer is specific to a generator:

1. For example, let's say one of our generators can be configured with custom attributes or naming convention. In that case we likely want an analyzers that can tell the developer when they messed up the configuration. Such an analyzer should likely ship in the same assembly as the generator because it's strongly tied.

2. A general purpose analyzer that is tied to a public API should likely ship like it does today because we generally don't need -- and don't want to -- tie those to a particular target framework.

Author: I was imagining an analyzer for a SQL connection string, or some other stringly typed API.  Any Analyzer trying to add better intellisense or design time validation of runtime functionality that's tied to features might make sense to ship this way.

Author: Clarified this a bit as

> Analyzers that need to validate source generator usage, or analyzers that are closely tied to runtime functionality may consider shipping side-by-side with the framework.

It may be desirable to create a separate package just for analyzers in order to reduce risk in servicing.  We support servicing ref packs today, but we have to be careful not to expose new API nor change any assembly identities.  In my opinion it makes more sense to invest efforts in ensuring safe servicing of ref packs, with potential increased complexity in our own builds.  There is benefit in having servicable ref packs regardless and this keeps additional complexity (resolution of additional pack) out of user builds.

### Why in the ref pack, why not runtime pack?

Runtime packs are currently independent of compile.  Sourcing inputs for compile from a runtime-specific package would break this and require significant changes to the user experience, such as permitting cross-compiling by runtime or "targeting" a runtime.

### Why in the ref pack, why not a dedicated NuGet package?

It's a goal for source-generators to be part of the default experience.  Additionally the default experience should not require a restore step.  As such analyzers cannot be delivered by NuGet packages alone.

> NuGet packages are a good option in addition to the ref pack for components which already ship as NuGet packages, eg System.Text.Json.  In this way the source generator can deliver the integrated experience of being *part of* the library.

Reviewer 10: ```suggestion
NuGet packages are a good option in addition to the ref pack for components which already ship as NuGet packages, such as `System.Text.Json`.  In this way the source generator can deliver the integrated experience of being *part of* the library.
```

> Standalone source-generator NuGet packages may be used for components which are in the shared framework when the source-generator is not ready to stabalize at the same time as the shared framework, and the component inside the shared framework does not need to be changed.  If the component in the shared framework needs to be changed, it should be done so by either a library+generator NuGet package (eg: System.Text.Json) or an experimental build of the entire framework, typically from a codebase like https://github.com/dotnet/runtimelab.

Reviewer 10: ```suggestion
Standalone source-generator NuGet packages may be used for components which are in the shared framework when the source-generator is not ready to stabalize at the same time as the shared framework, and the component inside the shared framework does not need to be changed.  If the component in the shared framework needs to be changed, it should be done so by either a library+generator NuGet package (for example `System.Text.Json`) or an experimental build of the entire framework, typically from a codebase like [dotnet/runtimelab](https://github.com/dotnet/runtimelab).
```

### Why in the ref pack, why not runtime pack?

Runtime packs are currently independent of compile.  Sourcing inputs for compile from a runtime-specific package would break this and require significant changes to the user experience, such as permitting cross-compiling by runtime or "targeting" a runtime.

### Why in the ref pack, why not a dedicated NuGet package?

It is a goal for source-generators to be part of the default experience.  Additionally, the default experience should not require a restore step.  As such analyzers cannot be delivered by NuGet packages alone.

NuGet packages are a good option in addition to the ref pack for components which already ship as NuGet packages, such as `System.Text.Json`.  In this way the source generator can deliver the integrated experience of being *part of* the library.

> Standalone source-generator NuGet packages may be used for components which are in the shared framework when the source-generator is not ready to stabalize at the same time as the shared framework, and the component inside the shared framework does not need to be changed.  If the component in the shared framework needs to be changed, it should be done so by either a library+generator NuGet package (`System.Text.Json`) or an experimental build of the entire framework, typically from a codebase like [dotnet/runtimelab](https://github.com/dotnet/runtimelab).

Reviewer 3: ```suggestion
Standalone source-generator NuGet packages may be used for components which are in the shared framework when the source-generator is not ready to stabilize at the same time as the shared framework, and the component inside the shared framework does not need to be changed.  If the component in the shared framework needs to be changed, it should be done so by either a library+generator NuGet package (`System.Text.Json`) or an experimental build of the entire framework, typically from a codebase like [dotnet/runtimelab](https://github.com/dotnet/runtimelab).
```

> Existing public functionality which controls the behavior of shared-framework references should also apply to source generators.  `$(DisableImplicitFrameworkReferences)` should disable the creation of `@(Analyzer)` items.  Any other property which might disable the creation of `@(Reference)` items from the ref-pack should also disable the creation of `@(Analyzer)` items.  Additionally, a new property should be introduced to disable the creation of `@(Analyzer)` items: `$(DisableFrameworkReferenceAnalyzers)`.

Author: Analyzer items need to point to the full path to the DLL.  It's directly passed to the compiler's -analyzer argument.

I don't think we need any more granular resolution here, we don't have such a feature for references.  It's all or nothing. 

If for some reason folks wanted to disable specific analyzers they could intercept the item in an msbuild target.  They would have to do the same for references today.

> Existing public functionality which controls the behavior of shared-framework references should also apply to source generators.  `$(DisableImplicitFrameworkReferences)` should disable the creation of `@(Analyzer)` items.  Any other property which might disable the creation of `@(Reference)` items from the ref-pack should also disable the creation of `@(Analyzer)` items.  Additionally, a new property should be introduced to disable the creation of `@(Analyzer)` items: `$(DisableFrameworkReferenceAnalyzers)`.

Author: I'll clarify this.

Source generators should not generate code which uses public API with "reserved" or "undocumented" functionality in attempts to avoid public API guarantees.  Libraries paired with source generators need to provide the same guarantees as libraries alone do today.

Source generators should avoid emitting public API in a user's projects.  Such API can create a compatibility contract in that user's library which needs to be maintained.  If a source generator produces public API in a user's project by design, that public API must be strictly maintained following the same rules as framework components across versions of the source generator.  Care should be taken to the public API which is generated and should be reviewed by the Framework Design Council.  A similar constraint applies to internal API: users will have source that expects to interact with the generated internal API and the user will expect that code to continue to compile between versions.

> Generated code should not be generated in System namespaces, it should prefer the user's namespace.  Generated code should be triple-slashed commented and be free from compiler warnings.  Generated code should avoid name collisions with user code and provide affordances for resolving name conflicts.  As much as possible generated code should meet the quality standards of .NET library code, using nullable annotations, meeting framework design guidelines, naming guidelines, etc.  It should feel like a natural extension of the library.

Author: @cartermp do we have guidance like this we'd prescribe all generators follow?  Should folks use `// <auto-generated>`  I was hoping to find this in https://github.com/dotnet/roslyn/blob/main/docs/features/source-generators.cookbook.md

> Generated code should not be generated in System namespaces, it should prefer the user's namespace.  Generated code should be triple-slashed commented and be free from compiler warnings.  Generated code should avoid name collisions with user code and provide affordances for resolving name conflicts.  As much as possible generated code should meet the quality standards of .NET library code, using nullable annotations, meeting framework design guidelines, naming guidelines, etc.  It should feel like a natural extension of the library.

Author: Perhaps more appropriately `GeneratedCodeAttribute` on types/members.

> Generated code should not be generated in System namespaces, it should prefer the user's namespace.  Generated code should be triple-slashed commented and be free from compiler warnings.  Generated code should avoid name collisions with user code and provide affordances for resolving name conflicts.  As much as possible generated code should meet the quality standards of .NET library code, using nullable annotations, meeting framework design guidelines, naming guidelines, etc.  It should feel like a natural extension of the library.

Reviewer 12: We should add that to the cookbook. Generally we advise authors to add `// <autogenerated>` at the top of the file, as it'll kick in the analyzer mechanism where it won't call back for generated code unless explicitly asked to. 

We do have an issue tracking to make this default however: generated files will *always* be treated as `<autogenerated>` by analyzers regardless of the annotation. https://github.com/dotnet/roslyn/issues/49326

> Generated code should not be generated in System namespaces, it should prefer the user's namespace.  Generated code should be triple-slashed commented and be free from compiler warnings.  Generated code should avoid name collisions with user code and provide affordances for resolving name conflicts.  As much as possible generated code should meet the quality standards of .NET library code, using nullable annotations, meeting framework design guidelines, naming guidelines, etc.  It should feel like a natural extension of the library.

Reviewer 12: I would guess you would want attributes too though, so that downstream tooling knows it was generated.

> Generated code should not be generated in System namespaces, it should prefer the user's namespace.  Generated code should be triple-slashed commented and be free from compiler warnings.  Generated code should avoid name collisions with user code and provide affordances for resolving name conflicts.  As much as possible generated code should meet the quality standards of .NET library code, using nullable annotations, meeting framework design guidelines, naming guidelines, etc.  It should feel like a natural extension of the library.

Author: Would it be possible for roslyn to automatically apply the `GeneratedCodeAttribute` to all things (types, or members of partial types) define by the generator?

> Generated code should not be generated in System namespaces, it should prefer the user's namespace.  Generated code should be triple-slashed commented and be free from compiler warnings.  Generated code should avoid name collisions with user code and provide affordances for resolving name conflicts.  As much as possible generated code should meet the quality standards of .NET library code, using nullable annotations, meeting framework design guidelines, naming guidelines, etc.  It should feel like a natural extension of the library.

Reviewer 13: @ericstj 

> Would it be possible for roslyn to automatically apply the GeneratedCodeAttribute to all things (types, or members of partial types) defined by the generator?

That doesn't really make sense at the Roslyn layer. These attributes change the behavior of code in a number of environments and it's likely not universally desired by all source generator authors. Instead it makes more sense as a helper to make it easier for generator authors with these need to more easily generate it.

> Generated code should not be generated in System namespaces, it should prefer the user's namespace.  Generated code should be triple-slashed commented and be free from compiler warnings.  Generated code should avoid name collisions with user code and provide affordances for resolving name conflicts.  As much as possible generated code should meet the quality standards of .NET library code, using nullable annotations, meeting framework design guidelines, naming guidelines, etc.  It should feel like a natural extension of the library.

Author: I see, we can do that for sure.  I suppose if it becomes prevalent and something that every generator ends up doing we could always propose a feature at that point.

### Source generator servicing

Source generators contribute significant code to user assemblies and that code may have issues.  We have very little control over changing customer assemblies once they have consumed a source generator.  We cannot and should not attempt to change the behavior of customer's assemblies (eg: hot patch) to fix issues.  Should an issue be found we should fix it in the source generator and ship a new version, just as we would if we found an issue with compiler.  Customers desiring this fix must proactively acquire the new code-generator, typically through an updated SDK which carries the ref pack, rebuild their code, and update their production applications.

Due to the limited servicing agility for code-generated code, we should be very careful about the complexity of the code which is generated.  Where possible we should limit generated code to "glue code", only that which must be specific to the user's assembly or types.  We should try to put more complex code inside the framework or library itself, even if it means exposing public API that is specifically for source generated code to call.

> Inbox source generators will be serviced in their side-by-side location.  Should a bug exist in a osurce generator we will evaluate the bug against the bar for each framework it exists and patch them independently.  In this way the servicing process for these source generators will match the .NET runtime.

Reviewer 14: Typo: `osurce`

Author: ```suggestion
Inbox source generators will be serviced in their side-by-side location.  Should a bug exist in a source generator we will evaluate the bug against the bar for each framework it exists and patch them independently.  In this way the servicing process for these source generators will match the .NET runtime.
```

> Inbox source generators will be serviced in their side-by-side location.  Should a bug exist in a source generator we will evaluate the bug against the bar for each framework it exists and patch them independently.  In this way the servicing process for these source generators will match the .NET runtime.

Author: The analyzer will be in it's own location.  Side-by-side in this context is talking about versions of the analyzer specific to each framework, rather than a central version used for all frameworks.  I can avoid the term `side-by-side` here if it creates confusion.  It's a bit of a relic of how we talked about .NETFramework components as side-by-side vs shared.

> Inbox source generators will be serviced in their side-by-side location.  Should a bug exist in a source generator we will evaluate the bug against the bar for each framework it exists and patch them independently.  In this way the servicing process for these source generators will match the .NET runtime.

Author: Scrubbed the document for uses of side by side and clarified them.

If a specific compiler version has a code-gen bug it is not clear from the final assembly if that bug is present, unless you directly inspect the IL/bytecode for the issue.  The compiler version is embedded in the PDB of the assembly which could be used to identify assemblies with the issue.  There is no notification at design time or runtime of any issue with the compiler, the user needs to subscribe to updates to understand when an update is needed.

If a runtime assembly has a bug, we have multiple avenues of detection.  We could examine the assembly metadata files contained in the application layout, including for self-contained applications.  The runtime will also write "breadcrumb" files on a machine when code is executed that is annotated for servicing that help identify that a specific assembly or package version has been executed on that machine.  Today we use these breadcrumbs to help target updates to a machine.  Those updates can install in a well-known location and be preferred over the application's copy.

For means of detection we have:
1. Manual
2. PDB indicator: source generator assembly version, file version, hash

> 3. Assembly metadata in consuming user project: GeneratedCodeAttribute("System.Foo.Generator", "5.0.0+cf258a14b70ad9069470a108f13765e0e5988f51")

Author: Actually this might be a misuse to place at the assembly level.  This attribute is meant to be inclusive, so it should only be placed on the type/members that are generated.

Reviewer 3: For detection purposes, it is enough to place it at assembly level.

Author: Right, but probably not this specific attribute which has existing meaning.  I'm making an update to the doc to make this distinction.

4. Runtime breadcrumbs

For means of update we have:
1. No notification, user needs to subscribe to servicing updates.
2. Design time notification.

> 3. Killbit in the runtime to prevent execution of user assemblies that contain vulnerable source generated code.

Reviewer 14: If we want to implement this one it would need the GeneratedCodeAttribute in the assembly metadata, right? PDB is generally not available to the runtime.

Author: Yeah.  We could impose this requirement ourselves in our source generators.  I think we need to understand if it's worth it to add a stamp to the runtime assembly or not.

Author: Discussion is happening here https://github.com/dotnet/designs/pull/181#discussion_r588772008

## Q & A

### Why not just deliver source generators in the SDK same as roslyn analyzers?

> Doing so would require source generators to carefully version and test across all framework versions.  This would increase testing cost for source generators and increase risk when shipping updates to source generators since they would apply to past, stable target frameworks.  Conceptually source generators are lifting a substantial amount of runtime functionality out of a library and executing that at design time. This code is directly included in the user's application and contributes to runtime behavior.

Author: That's a good point.  This will be a new requirement for framework components and it should be called out.

What this strategy does do is reduce risk.  We eliminate the channel where the latest SDK would contain changes that impacted users' projects targeting older .NET versions.  That is a good thing as it reduces the need to test those configurations and ensures we can garuntee the quality and stability of our stable .NET releases.

### Is it bad to have multiple versions of a source generator possibly loaded by a repository?

A repository may have projects targeting multiple frameworks.  These frameworks may contain different versions of the same source generator.  When building the repository this may require the same compiler server instance to load multiple copies of a source generator.  The same compiler would not likely load multiple copies of analyzers which do not ship side by side.  

### Should all inbox source generators be tied to a shared framework?

> No, some source generators might be independent of libraries in a shared framework: these could go in a non-side-by-side location (eg: analyzers folder).  Other source generators may not contain functionality but may instead just build some convenience API on top of existing public API.  These could be considered stable enough to place in a non-side-by-side location.

Author: ref-packs are specifically tied to framework versions.  There is no place in a ref-pack that is not tied to a framework version.

If we decide we want a source generator that is not tied to framework version I would expect it to go in a similar place to analyzers.

> No, some source generators might be independent of libraries in a shared framework: these could go in a non-side-by-side location (eg: analyzers folder).  Other source generators may not contain functionality but may instead just build some convenience API on top of existing public API.  These could be considered stable enough to place in a non-side-by-side location.

Reviewer 15: Sorry, I think my brain is not working with relation to this paragraph. Is this just saying that there exist scenarios for non-framework-specific in-box source generators, but they would not go through the deployment mechanisms described in this document for framework-specific in-box source generators?

> No, some source generators might be independent of libraries in a shared framework: these could go in a non-side-by-side location (eg: analyzers folder).  Other source generators may not contain functionality but may instead just build some convenience API on top of existing public API.  These could be considered stable enough to place in a non-side-by-side location.

Author: Yes.  There are likely source generators that are inbox, and maybe even built by the runtime or framework team, that don't need to ship as framework-specific.

We should be careful about such source generators though, as they would need to to be serviced in the same manner as the compiler: latest version needs to work everywhere and not break older apps using new SDK.

> No, some source generators might be independent of libraries in a shared framework: these could go in a non-side-by-side location (eg: analyzers folder).  Other source generators may not contain functionality but may instead just build some convenience API on top of existing public API.  These could be considered stable enough to place in a non-side-by-side location.

Author: I'll try to clarify this wording

It may be desirable to create a separate package just for analyzers to reduce risk in servicing.  We support servicing ref packs today, but we must be careful not to expose new API nor change any assembly identities.  In my opinion it makes more sense to invest efforts in ensuring safe servicing of ref packs, with potential increased complexity in our own builds.  There is benefit in having servicable ref packs regardless and this keeps additional complexity (resolution of additional pack) out of user builds.

### Why in the ref pack, why not runtime pack?

Runtime packs are currently independent of compile.  Sourcing inputs for compile from a runtime-specific package would break this and require significant changes to the user experience, such as permitting cross-compiling by runtime or "targeting" a runtime.

### Why in the ref pack, why not a dedicated NuGet package?

> It is a goal for source-generators to be part of the default experience.  Additionally, the default experience should not require a restore step.  As such analyzers cannot be delivered by NuGet packages alone.

Author: This is a good question.  I wasn't considering it within the scope of this document, but perhaps it should be.  I don't want to dig into details around how to make generators performant, that's more roslyn's territory.

We'll want to cross-gen them in the SDK installer like other design time code.  They need not be cross-gen'ed in the ref-pack nupkg since that is arch agnostic.

@chsienki do you foresee any other requirements around distribution that would impact the performance?  Will it be a problem if we have source-generator DLLs that are 1:1 with libraries?  

Should inbox source generators share any common infrastructure, APIs, or state?  Or perhaps if we find that such things are needed we push the functionality down into the roslyn APIs.

> It is a goal for source-generators to be part of the default experience.  Additionally, the default experience should not require a restore step.  As such analyzers cannot be delivered by NuGet packages alone.

Reviewer 13: We don't forsee in meaningful impact from source generators. We have a lot of experience with analyzers that stress the compiler in similar ways and our own experiments with generators don't show them to contribute meaningfully to the performance. In virtually every case where existing generators are replaced with roslyn source generators we are seeing build improvements (not the case here, but worth calling out). 

The biggest issue we have with generator and analyzer performance is bugs in the analyzers / generators themselves. It's not uncommon for us to get 2-3 bugs per VS release that we end up tracking down to misbehaving generators / analyzers. In all cases we work with those teams for a patch and get it quickly released to the community.

> It is a goal for source-generators to be part of the default experience.  Additionally, the default experience should not require a restore step.  As such analyzers cannot be delivered by NuGet packages alone.

Author: I was thinking more specifically about a design concern around number of modules.  Suppose we have 5 generators split into 5 modules, vs 1 module for all.  Do you see this as causing issues?  (Even trivial ones like some perf-gate somewhere notices and slows down a VS insertion).

> It is a goal for source-generators to be part of the default experience.  Additionally, the default experience should not require a restore step.  As such analyzers cannot be delivered by NuGet packages alone.

Author: Based on offline conversations we don't think this should be an issue.

# Inbox Source Generators

**PM** [Immo Landwerth](https://github.com/terrajobst) |
**Dev** [Eric StJohn](https://github.com/ericstj)

> Roslyn source generators allow for component developers to provide functionality that runs in the context of compilation, analyzes a user's code, and can generate additional code to add to the compilation.  Compelling scenarios for source generators involve moving dynamic runtime code to be specialized and statically emitted by the compiler: specializing serializer code for a target type, pre-compiling regular expressions at design time, pre-computing composition graphs for DI systems, and many more.  This technique was possible before but was rather clunky and had to be done outside the compiler, so its use was minimal and not part of the default experience.  Roslyn source generators solve this and as a result we wish to develop many source generators which are part of the default experience for .NET applications.  Source generators present a new type of challenge for a design time component: they need to be part of the toolchain but are tied to specific components within a target framework, and often re-implement or share the functionality of those components.  We need to define how we ship these source generators and make them available to developers.  We need to define the lifecycle for these source generators and how it is tied to the framework component it extends.

Reviewer 13: You mention "Roslyn source generators" which include both C# and VB generators yet the runtime is only going to include support for C#. May want to make that a bit clearer.

Reviewer 16: Also what does this mean to F# which is not Roslyn based? Will the user experience when using F# be comparatively worse compared to C#?

Author: > yet the runtime is only going to include support for C#

Is this a roslyn decision or one you're imagining we'll make in runtime?  I didn't go into those details here, but I'd expect us to do our best to support all languages possible, similar to analyzers, assuming the compiler supported them.  I have seen samples like @sharwell's that [include VB support](https://github.com/dotnet/roslyn-analyzers/pull/4897/files#diff-fe93da4b6169186079b6aabe3fafe56d70b0f47d0da01e6a825464bfaf1eb20c).

> Will the user experience when using F# be comparatively worse compared to C#

User experience of one language vs others is an interesting one.  I don't expect to cover that here.  This document assumes that the decision has been made to include a generator in the framework.  This could be something discussed for individual generators as they are proposed to the API review board, just like they consider API that is language specific.  FWIW we do spell out that the generator-produced code needs to rely exclusively on public API that meets normal public API compatibility criteria.  This should enable any other component to generate language specific sources to drive any API that supports source-generators.

cc @terrajobst

Reviewer 17: >assuming the compiler supported them

This is one of the limitations impacting the `DllImport` source generator - function pointers. This means VB will not be supported by the interop ones.

## Scenarios and User Experience

> A developer can use source generators for framework API without making any changes to their project.  Creating a new project should be sufficient.

Reviewer 13: These sentences seem to be in conflict: "without making any changes to their project" and "creating a new project should be sufficient"

Author: I'll clarify the second sentence.  It was meant to indicate that the user need not make any modifications.  It should say "targeting the framework that exposes the generator should be sufficient."

## Design

### Framework provided source generators

Framework provided source generators should have a mechanism to be exposed and versioned along side the framework and library which they extend.  These source generators will be passed to the compiler when that framework is targeted.  These source generators will not be used when the framework is not targeted.  For example, ASP.NET specific source generators will not be used when ASP.NET is not referenced.  The version of the source generators specific to the TargetFramework version will be used.  For example, when targeting .NET 6.0 the source generators specific to net6.0 will be used and not those for .NET 7.0, even when using the SDK that ships with .NET 7.0.    

Source generators and analyzers are represented and communicated in the same manner to the compiler, as analyzer dlls using the `-analyzer` switch.  Though this document is motivated by describing the process for source generators it will work in the same manner for analyzers.  In some places we may refer to the assemblies containing source generators as analyzers due to the way they are understood by the compiler, package manager, and SDK.

> To facilitate this, we will package source generators inside [reference packs](https://github.com/dotnet/designs/blob/main/accepted/2019/targeting-packs-and-runtime-packs.md).  We will use the standard [NuGet conventions for describing analyzers](https://docs.microsoft.com/en-us/nuget/guides/analyzers-conventions) with respect to the location in the package, however the SDK will not probe this path.  All source generators and their metadata will be listed in FrameworkList.xml, similar to references.  Analyzers will be added as `File` elements with `Type="Analyzer"` (constrasting to `Type="Managed"` which is currently used for references) and optionally `Language="cs|vb"`.  Omitting the `Language` attribute means the analyzer applies to all languages.  The .NET SDK will be responsible for locating the appropriate source generators from the refpack, based on the project's `$(Language)` and creating `@(Analyzer)` items as part of the `ResolveTargetingPackAssets` task.  These items will be conflict-resolved with other `@(Analyzer)` sources as described in the following section.  When viewed from the IDE it should be clear that the source of the `Analyzer` is the framework reference, similarly to how it is clear that the source of NuGet package analyzers come from packaages.

Reviewer 17: >`Language="cs|vb"`

I believe F# has some other mechanism for source generation but assume the current source generators will support it at some point. What is our extension story here for supporting future languages?

Reviewer 17: Basically who defines the new `Language` definition?

Author: This is just a parallel representation of the conventions supported by packages already. https://docs.microsoft.com/en-us/nuget/guides/analyzers-conventions

We're not defining anything new and if an analyzer were to support some new convention it should work the same way.  The SDK would need to handle any new analyzers in both places, likely by sharing this code: https://github.com/dotnet/sdk/blob/41cd7eb0d418f558c03459dfabef3601019d97b6/src/Tasks/Microsoft.NET.Build.Tasks/NuGetUtils.NuGet.cs#L50-L81

Author: I'll make it clear in the doc the values here should match

Existing public functionality which controls the behavior of shared-framework references should also apply to source generators.  `$(DisableImplicitFrameworkReferences)` should disable the creation of `@(Analyzer)` items.  Any other property which might disable the creation of `@(Reference)` items from the ref-pack should also disable the creation of `@(Analyzer)` items.  Additionally, a new property should be introduced to disable the creation of `@(Analyzer)` items: `$(DisableFrameworkReferenceAnalyzers)`.  No mechanism will be provided to select individual `@(Analyzer)` items, just as there is no mechanism to select individual `@(Reference)` items.

Alternative designs are listed in the [Q & A](#Q%20&%20A) section below.

### Source generator conflict resolution

Source generators and analyzers are passed to the compiler as `@(Analyzer)` items.  These items may come from NuGet packages, be built into the SDK targets, be directly defined by the user via `ProjectReference` or raw `Analyzer` items, or be provided by the new framework mechanism described in this document.  All should be considered for conflict resolution based on the same constraint.

> The .NET SDK already does conflict resolution for other asset types: references and copy-local runtime assets.  It should do the same for source generators, or more generally Analyzers.  Analyzers need not have "platform manifest" inputs for conflict resolution since all analyzers will be available during build, unlike runtime files which may not be available.  Comparison for the sake of conflict resolution can consider the same assembly and file version rules as reference files.  When finding two or more source generators with the same name conflict resolution will select the generator with higher assembly version, then higher file version, then that which is in the framework.  This parallels the same rules followed by reference assemblies.  The SDK may choose to optimize this comparison as it does for reference assemblies, by including a data file that includes analyzer metadata.

Reviewer 13: ```suggestion
The .NET SDK already does conflict resolution for other asset types: references and copy-local runtime assets.  It should do the same for source generators, or more generally analyzers.  Analyzers need not have "platform manifest" inputs for conflict resolution since all analyzers will be available during build, unlike runtime files which may not be available.  Comparison for the sake of conflict resolution can consider the same assembly and file version rules as reference files.  When finding two or more source generators with the same name conflict resolution will select the generator with higher assembly version, then higher file version, then that which is in the framework.  This parallels the same rules followed by reference assemblies.  The SDK may choose to optimize this comparison as it does for reference assemblies, by including a data file that includes analyzer metadata.
```

Consistency with other uses in the doc

Source generator conflict resolution should run before compile, but after all NuGet package assets have been evaluated.  Specific sequencing, public/private target naming, extensibility will not be specified here but should follow best practice as determined by the SDK team. 

### Source generator versioning

Source generators should be strongly named and update assembly version with every public stable release (GA and servicing).  This ensures that the publicly shipped source generators can be loaded side-by-side in the same process and follows our versioning guidelines for components which ship and must run on .NET Framework.  Source generators need not update assembly version during pre-release but should update file version to facilitate well behaved distribution by installers.  In general source generators need to follow all the same guidelines as other code that ships out of dotnet/runtime and runs outside the shared framework.

### Source generator compatibility concerns

> Source generators should only generate code which uses public API in framework components they extend.  This public API already needs to meet the strict binary compatibility guarantees as framework API.  As such libraries containing source-generated code will be binary-compatible with future releases, just as if the user wrote this code directly.

Reviewer 17: >As such libraries containing source-generated code will be binary-compatible with future releases, just as if the user wrote this code directly.

This does impose a non-obvious breaking change burden in the event we make breaking changes. We break rarely but now that we are following semantic versioning here, major versions can break when appropriate. Thoughts on how to track this? Are facilities needed to fail in predicable ways to help people understand these breaking changes.

Author: @terrajobst @jeffhandley 
This burden is the same we already impose on framework API, there's nothing new being imposed here.  This is just reminding folks that source-generators aren't some sort of loophole that's somehow less compatible.  If you have a source generator that you imagine isn't ready to meet those requirements then it probably shouldn't be put inbox, just like we don't put framework API in the box until we're certain it can meet those compatibility requirements.

Sure we may get things wrong and need to [obsolete](https://github.com/dotnet/designs/blob/9a6977bc6909049320695932d1a74a3a13c37a35/accepted/2020/better-obsoletion/better-obsoletion.md) API used by the generator.  I'm not seeing source generators having much of a difference in how this would work: we don't dig into binaries that aren't rebuilt to find obsolete APIs though we could and it would find source generated usage same as direct usage.  We can also obsolete the "entry point" API that source uses to interact with the generator.  If we were to do such a thing we might want to decide if the source-generator immediately stopped generating code in response to the obsoleted entry-point API or we phased it out like we do with framework API.  I think that'd need to be a context sensitive discussion that depends on the cost to the developer.

Generated code should not be generated in System namespaces, it should prefer the user's namespace.  Generated code should be triple-slashed commented and be free from compiler warnings.  Generated code should avoid name collisions with user code and provide affordances for resolving name conflicts.  As much as possible generated code should meet the quality standards of .NET library code, using nullable annotations, meeting framework design guidelines, naming guidelines, etc.  It should feel like a natural extension of the library.

### Deterministic failure

> The absence of a required source generator from code that requires it should produce a clear diagnosable error message from the compiler, like the error from a missing reference.  A codebase that was written to make use of a source-generator should never compile and silently omit the source generated code.  For example: if a source generator uses an attribute to activate the generator and then generates extension methods with overloads that are automatically preferred, the source generator should define the attribute in the source generator, not the framework library.  This will ensure that the project fails to compile if the source generator is absent.

Reviewer 17: >should define the attribute in the source generator, not the framework library

I think there is an outstanding GitHub issue on this and should be referenced here. This only works if the attribute in question has no inputs (i.e. is just an indication and doesn't contain additional options). The Roslyn source generator will indicate non-existing attributes as errors initially so only the fact that the attribute was emitted can be determined not any of its possible metadata.

Author: This document is not meant to call out all possible ways to achieve this behavior, nor link to point in time issues that might make achieving this behavior challenging.
The assertion here is that source-generators when explicitly addressed should be designed in such a way that the compiler will fail if the generator is absent.  
We should avoid the cases where we define framework API specifically for the source generator and then allow code to compile if the source generator is absent.

### Source generator servicing

Source generators contribute significant code to user assemblies and that code may have issues.  We have very little control over changing customer assemblies once they have consumed a source generator.  We cannot and should not attempt to change the behavior of customer's assemblies (eg: hot patch) to fix issues.  Should an issue be found we should fix it in the source generator and ship a new version, just as we would if we found an issue with compiler.  Customers desiring this fix must proactively acquire the new code-generator, typically through an updated SDK which carries the ref pack, rebuild their code, and update their production applications.

Due to the limited servicing agility for code-generated code, we should be very careful about the complexity of the code which is generated.  Where possible we should limit generated code to "glue code", only that which must be specific to the user's assembly or types.  We should try to put more complex code inside the framework or library itself, even if it means exposing public API that is specifically for source generated code to call.

> Inbox source generators will be serviced in their framework-specific location by shipping a new version of the ref-pack.  Should a bug exist in a source generator, we will evaluate the bug against the bar for each framework in which it exists and patch them independently.  In this way the servicing process for these source generators will match the .NET runtime.

Reviewer 13: This worries me. The servicing bar for the .NET runtime is very high compared to the .NET SDK which is very low. Generators exist more at the SDK layer and today SDK level servicing is relatively common. If we are putting Framework source generators where they require the .NET runtime servicing bar then I feel like we're going to be potentially setting ourselves up for pain here.

Author: The servicing bar for the compiler is even higher than the runtime though, right?
It's the same set of folks enforcing the bar and they can decide how to apply it.  Servicing changes for the SDK still need to go to tactics.  These components are emitting code into the user's assembly that impacts their application at runtime so I think it's correct to apply a higher bar than design-time-only SDK code.

Reviewer 13: > The servicing bar for the compiler is even higher than the runtime though, right?

No it's much lower.

When the runtime is serviced it is forced onto customers. For a given .NET release there is one runtime for the X different SDKs that we will produce on top of it. The compiler ships inside all these SDKs and can be free to be aggressive with changes because customers have to proactively consume the SDK. 

> Servicing changes for the SDK still need to go to tactics

There are two aspects to consider here:
1. How often the SDK changes for a given .NET runtime release 
2. The bar for servicing the SDK 

I'm talking about both here but primarily (1). The SDK changes *frequently* and with major revisions for a given release. It's not uncommon for this to reveal bugs / issues in existing analyzers / generators that need to be patched. Why do the runtime generators think that they won't have the same type of issues that we've seen with other components historically?

> These components are emitting code into the user's assembly that impacts their application at runtime so I think it's correct to apply a higher bar than design-time-only SDK code.

We have a fairly established bar for code changes in customer code. This feels like you're saying that existing bar (which we've used for many many years) is wrong. So I'd like to get more elaboration on this.

Author: > why do the runtime generators think that they won't have the same type of issues that we've seen with other components historically?

I don't think I was implying we wouldn't have to service these components or that they wouldn't have issues.  I'm imagining the majority of our servicing fixes for the generator are related to the generator's functionality, not it's interactions with the platform.  The source-generator platform is an app model where it needs to continue to work version-to-version, and the compiler would be just as eager to fix any compatibility bugs it introduced to break a source generator as it would be to tell all the source generators in the ecosystem to change to adopt a breaking change.  For the bad bugs we can always go back and service the generator in a particular framework's servicing branch, but for some it won't meet the bar and the "bug" would go away when the framework goes out of support.

>  This feels like you're saying that existing bar (which we've used for many many years) is wrong. So I'd like to get more elaboration on this.

I don't think I made any claims that the existing bar was wrong.  I'm saying a higher bar than **design-time only code**.  My impression is that generators should have a bar closer to framework code than the compiler due to the amount of loosely specified complexity they emit: small amount of user code results in large amount of complexity in the user's assembly.  I agree they are different than inbox framework code since they require SDK update and recompile, but probably require more scrutiny than OOB nuget packages (which also require a design time update) since the updates are bundled in the SDK as part of the default experience.  I tried to cover some of the differences here https://github.com/dotnet/designs/blob/e69303f5023302d6e60ac2220a0ff2d819b69609/accepted/2021/InboxSourceGenerators.md#inplace-servicing-updates-are-good-enough-for-the-compiler-why-not-source-generators

Another thing to consider that's related to this: most of the generators we're looking at now also ship with the libraries they extend in NuGet packages.  So we still need to drive fixes through dotnet/runtime to get those packages updated even if the generator also shipped framework-independent in the SDK.

Source generator assemblies execute in the compiler process and must target .NET Standard in order run on both .NET Framework and .NET Core. Source generators should support generating source for all target frameworks supported by their companion library.  In the case of inbox source generators discussed in this document this need only be the target framework the generator ships in.  In the case of a NuGet package provided source generator, this should be all frameworks supported by package.

Source generators should limit their dependencies.  There is no scheme for encoding framework-specific nor runtime-specific implementations of dependencies or source generators based on the runtime environment so source generators should avoid any dependencies on assemblies which need to differ by framework or runtime.  Source generators may depend on packages which are known dependencies of the compiler and should depend on versions less than or equal to that provided by the compiler and should not include that dependency in their package.

> Source generators should not rely on referencing nor executing framework types they wish to extend.  Despite inbox source generators being coupled to a specific framework version, source generators cannot assume they will be running on that framework.  They may be running on a newer, older, or completely different framework.  For example, API added to help a source generator do its job may not be present when running on .NET Framework or the .NET Core version the compiler is running on.  There is no type-equivalence guaranteed between types loaded in the compiler process and those of the user code that is under analysis.  For example, a source generator cannot instantiate attribute instances in the compiler process from user references to those attributes in source.  Should a source generator need to identify type usage it will need to examine user code references by name or resolve types in the user's references to ensure valid equivalence checks.  Source sharing should be considered as an alternative for cases where source generators need to execute runtime library helper code when generating source.  For example, serializer code which needs to examine the shape of user types can be shared between the runtime library and source generator.  Common abstractions can be made over metadata to allow the same code to handle different sources of metadata (source vs reflection).

Reviewer 13: > Despite inbox source generators being coupled to a specific framework version, source generators cannot assume they will be running on that framework

Can we make this language stronger? This sounds like the normal case is they run on the same framework and that seeing other frameworks is an uncommon event. The opposite is the truth. 

> Should a source generator need to identify type usage it will need to examine user code references by name or resolve types in the user's references to ensure valid equivalence checks

Can we reference the source generators cookbook here? It has examples on how to do this correctly.

> Source generators should localize any diagnostic messages displayed to the user.  This can be done using localized resource satellite assemblies.  These assemblies should be distributed with the source generator assembly. 

Reviewer 13: Can we have this document refer to [LocaliableResourceString](https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/Diagnostic/LocalizableResourceString.cs). This is the correct way to do localization in the compiler process. We don't use `CurrentCulture` bit instead use the culture specified on the compiler command line.  This type will take care of all the heavy lifting here and if it's not used then you're likely doing localization wrong 😄

## Q & A

### Why not just deliver source generators in the SDK same as roslyn analyzers?

> Doing so would require source generators to carefully version and test across all framework versions.  This would increase testing cost for source generators and increase risk when shipping updates to source generators since they would apply to past, stable target frameworks.  Conceptually source generators are lifting a substantial amount of runtime functionality out of a library and executing that at design time. This code is directly included in the user's application and contributes to runtime behavior.

Reviewer 13: Why do you believe that distributing the generators at the SDK level conflicts with the idea that they should be specific to the target framework? The build logic could easily pick the appropriate version of the generator for the target framework that is being built. Hence I'm unsure why you feel these ideas are in conflict.

Author: If the build logic is selecting the appropriate version of the generator than it effectively *is the same* as this proposal: we're just leveraging existing build logic for target framework selection by placing in the ref-pack rather than inventing something new.  This question is addressing the alternative of a single-non-target-framework-versioned copy of generators (as is done with analyzers)

### Can't the analyzer handle delivering different behavior per library version?

> We could implement the framework/library specific behavior in the analyzer itself, like the compiler does with `$(LangVersion)`.  We could use an input like the project's `$(TargetFramework)`, metadata in the reference asssembly like a source generator version experessed in an attribute `AssemblyMetadataAttribute("UseMySourceGeneratorVersion", "1.2")`, or even do API probing to determine which features are supported by the library.  Whichever input chosen would need to directly correlate with the beavior of the library selection (conflict resolution), ruling out `TargetFramework`, and be stable across servicing updates to provide determinsm to existing projects.  These solutions introduce complexity into the source generator and risk for the end user.  They would require us to build and test analyzers across all older targets: we would need to test the latest source generator across all previous versions of the library.  They also introduce risk since they could impact a customer's application when using the latest SDK but targeting an older, LTS framework, since those would use the newest analyzer undergoing churn that is not scrutinized through our normal servicing process.

Reviewer 17: ~~beavior~~ -> behavior

### Inplace servicing updates are good enough for the compiler, why not source generators?

The compiler has a very stable 1:1 relationship between a user's source and the code it generates. As such the user can control the output of the compiler through their own usage.  The compiler's contribution of complexity to the user's assembly is very small and highly specified.
 

> Contrast this to source generators which can take a very small amount of user source to produce a large amount of generated code.  This large amount of generated code contains significant complexity which is loosely specified.  Its behavior is specific to the component it extends.  It's likely that we'd need to bugfix this code over time and change it as we develop features in the components it extends.  It is also likely that we will learn new techniques and improvments to source generators over time, increasing the changes for bugfixes and features.

Reviewer 17: ~~improvments~~ -> improvements

> Users expect stability for projects which do not change, for example a project targeting an LTS release. It is much more difficult to ensure this stability when we need to significantly components which contribute complexity to that project.

Reviewer 17: > when we need to significantly components which contribute complexity to that project.

I am missing the concept here.

Author: This should say "It is much more difficult to ensure this stability when we need to significantly change components which contribute complexity to that project."

### Isn't this a lot of servicing if we find a bug?

Although this seems like more changes than if behavior was in the compiler-distributed analyzers it is not.  If we cared enough about a bug to service an analyzer that runs on a stable .NET version we would also want to service the SDK for that stable .NET Version: including the same number of changes and branches.  In fact, this strategy *reduces* servicing in the case where a bug only exists in a single version of source generator.

> What this strategy does do is reduce risk.  We eliminate the channel where the latest SDK would contain changes that impacted users' projects targeting older .NET versions.  That is a good thing as it reduces the need to test those configurations and ensures we can garuntee the quality and stability of our stable .NET releases.

Reviewer 17: ~~garuntee~~ -> guarantee

Author: Thanks.  I will never learn the correct spelling of this word.  🤦

### Is it bad to have multiple versions of a source generator possibly loaded by a repository?

> A repository may have projects targeting multiple frameworks.  These frameworks may contain different versions of the same source generator.  When building the repository this may require the same compiler instance to load multiple copies of a source generator.  The same compiler would not load multiple copies of source generators which ship in a non-framework specific location.  This should work and be supported based on source generator versioning.

Reviewer 13: Just want to add that it's already the case today that repositories load multiple different versions of analyzers. Don't see why generators would be any different here. Repositories and their NuGet references are complex and it's expected this can cause odd combinations of analyzers and generators to be used.

Author: Exactly.  I was trying to call this out, but also point out that we could do something wrong (eg: hardcoded assembly version) to break it, so we should be intentional about making this work.

### Should all inbox source generators be tied to a shared framework?

> No, some source generators might be independent of a target framework. One example of such a source generators may not contain functionality but may instead just build some convenience API on top of existing public API that is unlikely to ever change.  These could be considered stable enough to place in a non-framework-specific location like the SDK's analyzer folder.  Source generators which choose to ship this way should be highly specified and will need to support backwards compatibility and test on all frameworks.

Reviewer 18: ```suggestion
No, some source generators might be independent of a target framework. One example of such: a source generator may not contain functionality, but may instead build some convenience API on top of existing public API that is unlikely to ever change.  These could be considered stable enough to place in a non-framework-specific location like the SDK's analyzer folder.  Source generators which choose to ship this way should be highly specified and will need to support backwards compatibility and test on all frameworks.
```

### Should only source generators be shipped this way?  What about analyzers and code-fixes?

> Nothing is preventing analyzers and code fixes from shipping in the same manner.  If it is desirable for a such a component to ship with the framework rather than the compiler then that component may ship in this manner.  We should only do this if we have a good reason, stable analyzers that are lanaguage extensions or tied to stable public API should ship non-framework-specific with the SDK as they do today.  Analyzers that need to validate source generator usage, or analyzers that are closely tied to runtime functionality may consider shipping with the framework.

Reviewer 18: ```suggestion
Nothing is preventing analyzers and code fixes from shipping in the same manner.  If it is desirable for a such a component to ship with the framework rather than the compiler then that component may ship in this manner.  We should only do this if we have a good reason; stable analyzers that are language extensions or tied to stable public API should ship non-framework-specific with the SDK as they do today.  Analyzers that need to validate source generator usage, or analyzers that are closely tied to runtime functionality may consider shipping with the framework.
```

### Why in the ref pack, why not some other "analyzer" pack?

The SDK already has a means for acquisition and selection of ref packs.  The inputs to the selection of ref packs are the same as that of source generators (Version, FrameworkReference, and TargetFramework), and the outputs are consumed by the same phase (compile) with the same level of cross-targeting granularity (TargetFramework and references).

> It may be desirable to create a separate package just for analyzers to reduce risk in servicing.  We support servicing ref packs today, but we must be careful not to expose new API nor change any assembly identities.  In my opinion it makes more sense to invest efforts in ensuring safe servicing of ref packs, with potential increased complexity in our own builds.  There is benefit in having servicable ref packs regardless and this keeps additional complexity (resolution of additional pack) out of user builds.

Reviewer 18: ```suggestion
It may be desirable to create a separate package just for analyzers to reduce risk in servicing.  We support servicing ref packs today, but we must be careful not to expose new API nor change any assembly identities.  It makes more sense to invest efforts in ensuring safe servicing of ref packs, with potential increased complexity in our own builds.  There is benefit in having serviceable ref packs regardless and this keeps additional complexity (resolution of additional pack) out of user builds.
```

Standalone source-generator NuGet packages may be used for components which are in the shared framework when the source-generator is not ready to stabilize at the same time as the shared framework, and the component inside the shared framework does not need to be changed.  If the component in the shared framework needs to be changed, it should be done so by either a library+generator NuGet package (`System.Text.Json`) or an experimental build of the entire framework, typically from a codebase like [dotnet/runtimelab](https://github.com/dotnet/runtimelab).

###  I really really need to execute runtime code on the TargetFramework!

You probably need a designer or independent build component that has a host process that can leverage the runtime targeting information of the application.  At the moment this doesn't have a great pattern to follow with source generators.  The compiler does not provide a host, nor does it know anything about target runtime (or in some cases executable frameworks).

### How does this impact source build?

> For Linux distribution .NET defines a process from which the .NET product can be built from source: [dotnet/source-build](https://github.com/dotnet/source-build).  This process requires building the entire .NET product without any pre-built binaries.  Some components which make up .NET target older frameworks which aren't built from the latest codebase.  This is done because these components may need to run at design-time on a different framework (.NETFramework), or may also ship out-of-band as a NuGet package for use by libraries that work on many frameworks.  To support building these components on past frameworks, source-build uses an alternate representation of reference assets for those frameworks: [dotnet/source-build-reference-packages](https://github.com/dotnet/source-build-reference-packages).  In order to satisfy build-from-source packages should only contain reference-API and metadata, no executable code.  As a result, the source-build-reference-packages that represent past framework versions will omit source generators.  This could be a problem for any component which needs to build during source-build and targets this old framework, for example a project targeting .NET 6.0 in the .NET 7.0 source build.  That component would fail to build during source build due to the missing source-generator (see [Deterministic failure](#Deterministic%20failure)).  That component can mitigate the failure by either including a .NET 7.0 configuration for source build.  Alternatively, if the dependent source-generator is also out-of-band (for example: System.Text.Json) the failing component may reference the latest dependent out-of-band package to get the latest source-generator from the package instead of the ref-pack.  To force components to make this mitigation early, we will make sure that Arcade disables source-generators from ref-packs which are not the latest when building for source build.

Reviewer 18: ```suggestion
For Linux distribution .NET defines a process from which the .NET product can be built from source: [dotnet/source-build](https://github.com/dotnet/source-build).  This process requires building the entire .NET product without any pre-built binaries.  Some components which make up .NET target older frameworks which aren't built from the latest codebase.  This is done because these components may need to run at design-time on a different framework (.NETFramework), or may also ship out-of-band as a NuGet package for use by libraries that work on many frameworks.  To support building these components on past frameworks, source-build uses an alternate representation of reference assets for those frameworks: [dotnet/source-build-reference-packages](https://github.com/dotnet/source-build-reference-packages).  In order to satisfy build-from-source, packages should only contain reference-API and metadata, no executable code.  As a result, the source-build-reference-packages that represent past framework versions will omit source generators.  This could be a problem for any component which needs to build during source-build and targets this old framework, for example a project targeting .NET 6.0 in the .NET 7.0 source build.  That component would fail to build during source build due to the missing source-generator (see [Deterministic failure](#Deterministic%20failure)).  That component can mitigate the failure by either including a .NET 7.0 configuration for source build.  Alternatively, if the dependent source-generator is also out-of-band (for example: System.Text.Json) the failing component may reference the latest dependent out-of-band package to get the latest source-generator from the package instead of the ref-pack.  To force components to make this mitigation early, we will make sure that Arcade disables source-generators from ref-packs which are not the latest when building for source build.
```
